osquery/carver/carver.cpp

```diff
@@ -55,60 +54,74 @@ CLI_FLAG(uint32,
          8192,
          "Size of blocks used for POSTing data back to remote endpoints");
 
-CLI_FLAG(bool,
-         disable_carver,
-         true,
-         "Disable the osquery file carver (default true)");
-
-CLI_FLAG(bool,
-         carver_disable_function,
-         FLAGS_disable_carver,
-         "Disable the osquery file carver function (default true)");
-
 CLI_FLAG(bool,
          carver_compression,
          false,
          "Compress archives using zstd prior to upload (default false)");
 
+DECLARE_bool(disable_carver);
 DECLARE_uint64(read_max);
 
-/// Helper function to update values related to a carve
-void updateCarveValue(const std::string& guid,
-                      const std::string& key,
-                      const std::string& value) {
-  std::string carve;
-  auto s = getDatabaseValue(kCarveDbDomain, kCarverDBPrefix + guid, carve);
-  if (!s.ok()) {
-    VLOG(1) << "Failed to update status of carve in database " << guid;
-    return;
-  }
+std::atomic<bool> CarverRunnable::running_{false};
 
-  JSON tree;
-  s = tree.fromString(carve);
-  if (!s.ok()) {
-    VLOG(1) << "Failed to parse carve entries: " << s.what();
-    return;
-  }
+void CarverRunnable::start() {
+  std::vector<std::string> carves;
+  scanDatabaseKeys(kCarves, carves, kCarverDBPrefix);
 
-  tree.add(key, value);
+  for (const auto& key : carves) {
+    std::string carve;
+    auto s = getDatabaseValue(kCarves, key, carve);
+    if (!s.ok()) {
+      VLOG(1) << "Failed to retrieve carve key: " << key;
+      deleteDatabaseValue(kCarves, key);
+      continue;
+    }
 
-  std::string out;
-  s = tree.toString(out);
-  if (!s.ok()) {
-    VLOG(1) << "Failed to serialize carve entries: " << s.what();
-  }
+    JSON tree;
+    s = tree.fromString(carve);
+    if (!s.ok() || !tree.doc().IsObject()) {
+      deleteDatabaseValue(kCarves, key);
+      VLOG(1) << "Failed to parse carve entries: " << s.getMessage();
+      continue;
+    }
 
-  s = setDatabaseValue(kCarveDbDomain, kCarverDBPrefix + guid, out);
-  if (!s.ok()) {
-    VLOG(1) << "Failed to update status of carve in database " << guid;
+    auto& doc = tree.doc();
+    if (!doc.HasMember("status") || !doc["status"].IsString()) {
+      // Malformed data.
```

Comment:
If there's malformed carve entries in the DB, would it also be worth having another `VLOG` statement here?

Reply:
It would, I plan to have some followup commits that at least:
- change the flags to non-CLI flags
- unify how VLOGs / LOGs are used

Reply:
Were you still planning to add a `VLOG` in here? Or do you think we don't need one?